src/lem/constrainer.rs

```diff
@@ -81,145 +116,390 @@ impl LEM {
-        match concrete_path {
-            None => Ok(true),
-            Some(concrete_path) => concrete_path
-                .get_value()
-                .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),
-        }
-    }
-
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn on_concrete_path(concrete_path: &Boolean) -> Result<bool> {
+        concrete_path
+            .get_value()
+            .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path"))
+    }
+
+    fn z_ptr_from_valuation<F: LurkField>(
+        concrete_path: &Boolean,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(concrete_path)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
+    /// Accumulates slot data that will be used later to generate the constraints.
+    fn acc_slots_data_data<F: LurkField>(
+        concrete_path: Boolean,
+        slots_data: &mut Vec<SlotData<F>>,
+        img: AllocatedNum<F>,
+        preimg_vec: Vec<&AllocatedPtr<F>>,
+    ) {
+        let arity = match preimg_vec.len() {
+            2 => HashArity::A2,
+            3 => HashArity::A3,
+            4 => HashArity::A4,
+            _ => unreachable!(),
+        };
+        let mut preimg = Vec::with_capacity(preimg_vec.len() * 2);
+        for elem in preimg_vec {
+            preimg.push(elem.tag().clone());
+            preimg.push(elem.hash().clone());
+        }
+        slots_data.push(SlotData {
+            arity,
+            concrete_path,
+            preimg,
+            img,
+        });
+    }
+
+    /// Use the implies logic to constrain tag and hash values for accumulated
+    /// slot data
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
```

Comment:
Whimiscal question: if this is all the implementation of a constrainer(.rs), should this be `constrain_slot`?